### PR TITLE
ext. filtering for lang matching

### LINKED ISSUE #15746: Implement extended filtering for language matching
#15464 added support for the lang pseudo-class, but it is using an equality comparison to check the language of the element with the languages specified in `:lang()`. We should instead use the algorithm specified [here](https://tools.ietf.org/html/rfc4647#section-3.3.2) for matching languages.

Spec: https://tools.ietf.org/html/rfc4647#section-3.3.2
Files: `components/script/dom/element.rs`, `components/script/layout_wrapper.rs`
Tests: `tests/wpt/web-platform-tests/dom/nodes/Element-matches.html`, `tests/wpt/css-tests/selectors-3_dev/`, ` tests/wpt/web-platform-tests/html/dom/elements/global-attributes/` (use `./mach test-wpt` with the appropriate file path to run the tests)

### PR DESCRIPTION
**Pull request by sendilkumarn**
This adds the extended algorithm to check the lang. This is my very first encounter with`rust`.
Open to learn. 
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15746

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15797)

### CODE REVIEW

File: components/script/dom/element.rs [hunk 1]
```diff
@@ -2974,3 +2991,53 @@ pub fn cors_setting_for_element(element: &Element) -> Option<CorsSettings> {
         }
     })
 }
+
+pub fn lang_check(css_lang: &Box<str>, element_lang: String) -> bool {
```
Comment (KiChjang) [inline components/script/dom/element.rs]: This doesn't need to be here; it can live in `components/script/dom/bindings/str.rs`.

Also, the parameters can just be `tag: &str, range: &str`. It's also a good idea to link to https://tools.ietf.org/html/rfc4647#section-3.3.2 above the function to indicate that this is following a specification.

Finally, the function name can just be `extended_filtering`.

File: components/script/dom/element.rs [hunk 2]
```diff
@@ -2974,3 +2991,53 @@ pub fn cors_setting_for_element(element: &Element) -> Option<CorsSettings> {
...
+    let css_lang_tags: Vec<&str> = css_lang.split('\x2d').collect();
```
Comment (KiChjang) [inline components/script/dom/element.rs]: Looks wrong. The CSS lang attribute is a language priority list, which contains language ranges. @SimonSapin would that be a correct interpretation?

File: components/script/dom/element.rs [hunk 3]
```diff
@@ -2402,7 +2412,7 @@ impl<'a> ::selectors::Element for Root<Element> {
 
             // FIXME(#15746): This is wrong, we need to instead use extended filtering as per RFC4647
             //                https://tools.ietf.org/html/rfc4647#section-3.3.2
-            NonTSPseudoClass::Lang(ref lang) => lang.eq_ignore_ascii_case(&self.get_lang()),
+            NonTSPseudoClass::Lang(ref lang) => self.is_lang_matches(lang),
```
Comment (KiChjang) [inline components/script/dom/element.rs]: If you simply call `lang_check(&*self.get_lang(), &*lang)`, you can eliminate the `is_lang_matches` functions.

File: components/script/dom/element.rs [hunk 4]
```diff
@@ -2974,3 +2991,53 @@ pub fn cors_setting_for_element(element: &Element) -> Option<CorsSettings> {
...
+        }
+
+        if element_tag.eq_ignore_ascii_case(css_tag) {
+            css_tag_index += 1;
+            tag_index += 1;
+        } else {
+            is_match = false;
+            break;
+        }
+    }
```
Comment (KiChjang) [inline components/script/dom/element.rs]: This algorithm is very hard to follow. It also looks like step 2 of it is missing. Can you add in step annotations?

File: components/script/dom/element.rs [hunk 5]
```diff
@@ -2974,3 +2991,53 @@ pub fn cors_setting_for_element(element: &Element) -> Option<CorsSettings> {
...
+    return is_match;
```
Comment (KiChjang) [inline components/script/dom/element.rs]: No need for an explicit `return` keyword here; Rust automatically returns the last expression in a function.

File: tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-005.htm.ini [hunk 6]
```diff
@@ -1,5 +1,5 @@
 [css3-selectors-lang-005.htm]
   type: testharness
   [A :lang value will match a lang attribute value when the latter contains additional subtags.]
-    expected: FAIL
+    expected: PASS
```
Comment (KiChjang) [inline tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-005.htm.ini]: You can simply remove this file. By default, tests are expected to PASS, the presence of an ini file is to explicitly mark them as either FAIL or TIMEOUT.

**Review by KiChjang — CHANGES_REQUESTED** [review]

File: tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht [hunk 7]
```diff
@@ -27,7 +27,7 @@ This tests a detail related to :lang support. If :lang is not supported, a messa
 <script>
 test(function() {
 assert_equals(document.getElementById('colonlangcontroltest').offsetWidth, 0)
-assert_equals(document.getElementById('box').offsetWidth, 50);
+assert_equals(document.getElementById('box').offsetWidth, 100);
```
Comment (KiChjang) [inline tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht]: You can't just simply change the tests like that to fit your behavior. These are tests that are shared among all browsers. What is the actual problem here that makes you want to change the assertion?

File: tests/wpt/css-tests/selectors-3_dev/xhtml1print/css3-selectors-lang-014.xht [hunk 8]
```diff
@@ -27,7 +27,7 @@ This tests a detail related to :lang support. If :lang is not supported, a messa
 <script>
 test(function() {
 assert_equals(document.getElementById('colonlangcontroltest').offsetWidth, 0)
-assert_equals(document.getElementById('box').offsetWidth, 50);
+assert_equals(document.getElementById('box').offsetWidth, 100);
```
Comment (KiChjang) [inline tests/wpt/css-tests/selectors-3_dev/xhtml1print/css3-selectors-lang-014.xht]: Ditto here.

File: tests/wpt/metadata-css/selectors-3_dev/xhtml1/css3-selectors-lang-005.xht.ini [hunk 9]
```diff
@@ -1,5 +1,5 @@
 [css3-selectors-lang-005.xht]
   type: testharness
   [A :lang value will match a lang attribute value when the latter contains additional subtags.]
-    expected: FAIL
+    expected: PASS
```
Comment (KiChjang) [inline tests/wpt/metadata-css/selectors-3_dev/xhtml1/css3-selectors-lang-005.xht.ini]: This hasn't been addressed.

File: components/script/dom/bindings/str.rs [hunk 10]
```diff
@@ -113,6 +113,90 @@ pub fn is_token(s: &[u8]) -> bool {
...
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let range_tags: Vec<&str> = range.split(',').collect();
+    let mut is_match: bool = true;
+
+    for range_tag in range_tags {
+        // (step 1) Split both the extended language range and the language tag being
+        // compared into a list of subtags by dividing on the hyphen (%x2D)
+        // character.  Two subtags match if either they are the same when
+        // compared case-insensitively or the language range's subtag is the
+        // wildcard '*'.
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: A comment saying step 1 is enough, we don't need the full spec text here.

File: components/script/dom/bindings/str.rs [hunk 11]
```diff
@@ -113,6 +113,90 @@ pub fn is_token(s: &[u8]) -> bool {
...
+        let elem_tags: Vec<&str> = tag.split('\x2d').collect();
+        if tag.eq_ignore_ascii_case(range_tag) || range_tag.eq_ignore_ascii_case("*") {
+            return true;
+        }
+
+        // (step 2) Begin with the first subtag in each list.  If the first subtag in
+        // the range does not match the first subtag in the tag, the overall
+        // match fails.  Otherwise, move to the next subtag in both the
+        // range and the tag.
+        // Note: [Level-4 spec](https://drafts.csswg.org/selectors/#lang-pseudo) check for wild card
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Same for this and the other ones.

**Review by KiChjang — CHANGES_REQUESTED** [review]

File: tests/wpt/metadata-css/selectors-3_dev/xhtml1/css3-selectors-lang-005.xht.ini (hunk 9 above)
Comment (sendilkumarn) [inline tests/wpt/metadata-css/selectors-3_dev/xhtml1/css3-selectors-lang-005.xht.ini]: By spec, `es` should match `es-MX` so we cannot have a fail here. 
I even checked in browsers and it does work in the same way.

Comment (KiChjang) [inline tests/wpt/metadata-css/selectors-3_dev/xhtml1/css3-selectors-lang-005.xht.ini]: No, I mean that you can remove this file.

Comment (KiChjang) [inline tests/wpt/metadata-css/selectors-3_dev/xhtml1/css3-selectors-lang-005.xht.ini]: No, I mean that you can remove this file. And all the other ones that now pass.

File: tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht (hunk 7 above)
Comment (sendilkumarn) [inline tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht]: as the spec says, 
``` 
For example, the extended language range "de-*-DE" (or its synonym "de-DE") matches all of the following tags:
 de-de (German, as used in Germany)
de-Latn-DE (Latin script)
```
so it is the same test case over, which should succeed. But am not too sure on this since browsers implemented it differently.

Comment (sendilkumarn) [inline tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht]: [check example 28 here](https://drafts.csswg.org/selectors/#the-lang-pseudo)

File: tests/wpt/metadata-css/selectors-3_dev/xhtml1/css3-selectors-lang-005.xht.ini (hunk 9 above)
Comment (sendilkumarn) [inline tests/wpt/metadata-css/selectors-3_dev/xhtml1/css3-selectors-lang-005.xht.ini]: Got it. 👍 Thanks a lot again.

File: components/script/dom/bindings/str.rs [hunk 12]
```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
     })
 }
 
+/// Returns whether the language is matched, as defined by
+/// [RFC 4647](https://tools.ietf.org/html/rfc4647#section-3.3.2).
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let range_tags: Vec<&str> = range.split(',').collect();
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: It's not actually known what the separator is between different language ranges. I suspect it's the empty space character. @SimonSapin do you happen to know?

File: components/script/layout_wrapper.rs [hunk 13]
```diff
@@ -503,6 +503,7 @@ impl<'le> ServoLayoutElement<'le> {
 
         true
     }
+
```
Comment (KiChjang) [inline components/script/layout_wrapper.rs]: nit: Remove this newline.

**Comment by KiChjang** [comment]
I still need to review through the actual implementation of the algorithm. I'm not too convinced that it's currently at its cleanest, primarily because it's done using `for` loops. We can definitely try to utilize Rust's iterators there.

File: components/script/dom/bindings/str.rs (hunk 12 above)
Comment (sendilkumarn) [inline components/script/dom/bindings/str.rs]: > It accepts a comma-separated list of one or more language ranges as its argument.

I made a decision based on this.

**Comment by sendilkumarn** [comment]
@KiChjang yeah exactly, since it is my first time in rust more eager to learn and implement. Would love to get more comments.

File: tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht (hunk 7 above)
Comment (KiChjang) [inline tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht]: So one thing to note: this test file is under selectors 3, where wildcard matching is not implemented. In a sense, it's great that we fail this test, because that means we actually follow selectors 4.
In other words, please revert this change and mark this test as failing. Same for the other ones that you've changed under selectors 3.

File: components/script/dom/bindings/str.rs (hunk 12 above)
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: I see, that looks correct, but can you change this variable name to `lang_ranges` instead? A range is not a tag.

File: components/script/dom/bindings/str.rs [hunk 14]
```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
...
+    let mut is_match: bool = true;
+
+    for range_tag in range_tags {
+        // step 1
+        let css_tags: Vec<&str> = range_tag.split('\x2d').collect();
+        let elem_tags: Vec<&str> = tag.split('\x2d').collect();
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Let's not introduce the `css` and `elem` terms here. The extended filtering algorithm does not define these at all. Instead, name them `range_subtags` and `tag_subtags`.

File: components/script/dom/bindings/str.rs [hunk 15]
```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
...
+            continue;
+        }
+
+        // step 3
+        let css_tag_size = css_tags.len();
+        let elem_tag_size = elem_tags.len();
+        let mut css_tag_index = 1;
+        let mut elem_tag_index = 1;
+
+        while css_tag_size > css_tag_index {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Following on to the last one, instead of using a `while` loop here, loop through the rest of `range_iter`'s elements using `for` and perform the substeps A to E in the loop.

File: tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht (hunk 7 above)
Comment (sendilkumarn) [inline tests/wpt/css-tests/selectors-3_dev/xhtml1/css3-selectors-lang-014.xht]: only for `css3-selectors-lang-014` right

both `css3-selectors-lang-009 / 005` are fine

File: components/script/dom/bindings/str.rs [hunk 16]
```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
...
+            // step 3a
+            if css_tags[css_tag_index].eq_ignore_ascii_case("*") {
+                css_tag_index += 1;
+                continue;
+            }
+            // step 3b
+            if elem_tag_size <= elem_tag_index {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Using the new iterator method I proposed, this can adequately be implemented by just calling [`next`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#tymethod.next) on `tag_iter` and checking if the return value is `None`.

File: components/script/dom/bindings/str.rs [hunk 17]
```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
...
+        // step 1
+        let css_tags: Vec<&str> = lang_range.split('\x2d').collect();
+        let elem_tags: Vec<&str> = tag.split('\x2d').collect();
+        if tag.eq_ignore_ascii_case(lang_range) || lang_range.eq_ignore_ascii_case("*") {
+            return true;
+        }
+
+        // step 2
+        // Note: [Level-4 spec](https://drafts.csswg.org/selectors/#lang-pseudo) check for wild card
+        if !(css_tags[0].eq_ignore_ascii_case(elem_tags[0]) || css_tags[0].eq_ignore_ascii_case("*")) {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Let's not directly index the first element in the arrays. Instead, create an iterator on both of them, and use the following:
```rust
if let (Some(range_subtag), Some(tag_subtag)) = (range_iter.next(), tag_iter.next()) {
    // Do subtag matching...
}
```

File: components/script/dom/bindings/str.rs [hunk 18]
```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
...
+/// Returns whether the language is matched, as defined by
+/// [RFC 4647](https://tools.ietf.org/html/rfc4647#section-3.3.2).
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let lang_ranges: Vec<&str> = range.split(',').collect();
+    let mut is_match: bool = true;
+
+    for lang_range in lang_ranges {
+        // step 1
+        let css_tags: Vec<&str> = lang_range.split('\x2d').collect();
+        let elem_tags: Vec<&str> = tag.split('\x2d').collect();
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Let's not call these css and elem tags. RFC 4647 does not define these terms. Instead, name them `range_subtags` and `tag_subtags`.

File: components/script/dom/bindings/str.rs [hunk 19]
```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
     })
 }
 
+/// Returns whether the language is matched, as defined by
+/// [RFC 4647](https://tools.ietf.org/html/rfc4647#section-3.3.2).
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let lang_ranges: Vec<&str> = range.split(',').collect();
+    let mut is_match: bool = true;
+
+    for lang_range in lang_ranges {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Instead of using a `for` loop here, you can use [`any`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any), since the overall match succeeds if at least one of the range matches with the tag. You can also eliminate the `is_match` variable this way.

File: components/script/dom/bindings/str.rs [hunk 20]
```diff
@@ -113,6 +113,74 @@ pub fn is_token(s: &[u8]) -> bool {
...
+                    break;
+                } else {
+                    // step 3e
+                    elem_tag_index += 1;
+                }
+            }
+        }
+
+        // step 4
+        if css_tag_size <= css_tag_index && is_match {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: In the new iterator method I proposed, this would just need to return true.

**Review by KiChjang — CHANGES_REQUESTED** [review]

File: components/script/dom/bindings/str.rs [hunk 21]
```diff
@@ -113,6 +113,63 @@ pub fn is_token(s: &[u8]) -> bool {
...
+    lang_ranges.iter().any(|&lang_range| {
+        // step 1
+        let range_subtags: Vec<&str> = lang_range.split('\x2d').collect();
+        let tag_subtags: Vec<&str> = tag.split('\x2d').collect();
+        if tag.eq_ignore_ascii_case(lang_range) || lang_range.eq_ignore_ascii_case("*") {
+            return true;
+        }
+
+        let mut range_iter = range_subtags.iter();
+        let mut tag_iter = tag_subtags.iter();
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Why do these two have to be mutable?

File: components/script/dom/bindings/str.rs [hunk 22]
```diff
@@ -113,6 +113,63 @@ pub fn is_token(s: &[u8]) -> bool {
...
+/// [RFC 4647](https://tools.ietf.org/html/rfc4647#section-3.3.2).
+pub fn extended_filtering(tag: &str, range: &str) -> bool {
+    let lang_ranges: Vec<&str> = range.split(',').collect();
+
+    lang_ranges.iter().any(|&lang_range| {
+        // step 1
+        let range_subtags: Vec<&str> = lang_range.split('\x2d').collect();
+        let tag_subtags: Vec<&str> = tag.split('\x2d').collect();
+        if tag.eq_ignore_ascii_case(lang_range) || lang_range.eq_ignore_ascii_case("*") {
+            return true;
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: This check is unnecessary here. The spec at this point only defines what it means to have a match, it doesn't intend for you to actually make a match here.

File: components/script/dom/bindings/str.rs [hunk 23]
```diff
@@ -113,6 +113,63 @@ pub fn is_token(s: &[u8]) -> bool {
...
+            }
+        }
+
+        // step 3
+        for range_subtag in range_iter {
+            // step 3a
+            if range_subtag.eq_ignore_ascii_case("*") {
+                continue;
+            }
+            loop {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Looks wrong. "Continue with the loop" in substep E means the outer `for` loop, unless you're creating this `loop` for other reasons? I have no idea what you're trying to achieve here.

File: components/script/dom/bindings/str.rs [hunk 24]
```diff
@@ -113,6 +113,63 @@ pub fn is_token(s: &[u8]) -> bool {
...
+                        // step 3c
+                        if range_subtag.eq_ignore_ascii_case(tag_subtag) {
+                            break;
+                        } else {
+                            // step 3d
+                            if tag_subtag.len() == 1 {
+                                return false;
+                            } else {
+                                // step 3e
+                                tag_iter.next();
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Looks wrong. `tag_iter.next()` already advances the iterator to the next element, and here you're advancing it the second time.

File: components/script/dom/bindings/str.rs [hunk 25]
```diff
@@ -113,6 +113,63 @@ pub fn is_token(s: &[u8]) -> bool {
...
+                            }
+                        }
+                    },
+                    // step 3b
+                    None => { return false; }
+                }
+            }
+        }
+        // step 4
+        return true;
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: nit: No need for the return keyword here, Rust automatically treats the last expression as the return value.

**Review by KiChjang — CHANGES_REQUESTED** [review]
Almost there! I suspect that you will encounter more test expectation changes once this is properly solved.

File: components/script/dom/bindings/str.rs (hunk 23 above)
Comment (sendilkumarn) [inline components/script/dom/bindings/str.rs]: `substep E` only tells to go to `next` in `tag` not `range`, if we go with `for` loop then the `range` will go to `next`, which will have undesirable effects.

File: components/script/dom/bindings/str.rs (hunk 21 above)
Comment (sendilkumarn) [inline components/script/dom/bindings/str.rs]: `if let (Some(range_subtag), Some(tag_subtag)) = (range_iter.next(), tag_iter.next()) {` 
since we are borrowing it mutably here.

Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Why is this a mutable borrow?

File: components/script/dom/bindings/str.rs (hunk 23 above)
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Okay, so here's where we're actually going off spec a bit: the tag's current subtag is obtained by calling `next()`. Notice in the text that every time you move to the next subtag in the language tag, it is always followed by continuing the loop. Hence, when the spec says current, you can call next() to advance the iterator, and when the spec says move to the next subtag in tag, you should do nothing, since the next iteration will call next() automatically when calculating it's current subtag. Does that make sense?

Comment (sendilkumarn) [inline components/script/dom/bindings/str.rs]: this is interesting. 
This pretty much looks good. But the problem is when you have the tag length as `2` ( actually when range and tag is equal in length ).

consider this, 
tag -> `en-GB`
range -> `en-US`

if we don't have that inner loop. 
step 2 - tag and range will match. (`en`)

step 3 -  (first loop)
    a - not applicable
    b - not applicable
    c - will fail
    d - will fail
    e - will move to the next subtag. which will be `None`

Step 3 will finish here since there will not be any other elements in range

Step 4 - returns true ( which is wrong )

In other scenarios it does make sense. am I missing anything here.

Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: First, I do believe that this loop has to go away, regardless of which step we take next, because step 3e continues this loop, rather than the outer for loop.
So, 2 possible solutions here, either we duplicate substep A and B after the for loop, or we follow the spec strictly and set up some variables first before the for loop, so something like `let mut current_tag_subtag = tag_iter.next();` and we match against current_tag_subtag and updating it as necessary as outlined in the spec.

File: components/script/dom/bindings/str.rs [hunk 26]
```diff
@@ -113,6 +113,59 @@ pub fn is_token(s: &[u8]) -> bool {
...
+        // step 1
+        let range_subtags: Vec<&str> = lang_range.split('\x2d').collect();
+        let tag_subtags: Vec<&str> = tag.split('\x2d').collect();
+
+        let mut range_iter = range_subtags.iter();
+        let mut tag_iter = tag_subtags.iter();
+
+        // step 2
+        // Note: [Level-4 spec](https://drafts.csswg.org/selectors/#lang-pseudo) check for wild card
+        if let (Some(range_subtag), Some(tag_subtag)) = (range_iter.next(), tag_iter.next()) {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Why are these considered mutable borrows?

**Review by KiChjang — CHANGES_REQUESTED** [review]

File: components/script/dom/bindings/str.rs (hunk 23 above)
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Note that for the latter solution, you will most likely need to match against the clone of current_tag_subtag to avoid borrowck complaining.

File: components/script/dom/bindings/str.rs (hunk 26 above)
Comment (sendilkumarn) [inline components/script/dom/bindings/str.rs]: doesn't `next()` borrow mutably

Comment (sendilkumarn) [inline components/script/dom/bindings/str.rs]: (here)[https://play.rust-lang.org/?code=fn%20main()%20%7B%0Alet%20a%20%3D%20%5B1%2C%202%2C%203%5D%3B%0A%0Alet%20mut%20iter%20%3D%20a.iter()%3B%0A%0A%2F%2F%20A%20call%20to%20next()%20returns%20the%20next%20value...%0Aassert_eq!(Some(%261)%2C%20iter.next())%3B%0Aassert_eq!(Some(%262)%2C%20iter.next())%3B%0Aassert_eq!(Some(%263)%2C%20iter.next())%3B%0A%0A%2F%2F%20...%20and%20then%20None%20once%20it%27s%20over.%0Aassert_eq!(None%2C%20iter.next())%3B%0A%0A%2F%2F%20More%20calls%20may%20or%20may%20not%20return%20None.%20Here%2C%20they%20always%20will.%0Aassert_eq!(None%2C%20iter.next())%3B%0Aassert_eq!(None%2C%20iter.next())%3B%0A%7D]

the compilation will fail if I remove `mut`

Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Ah yes, that is correct. Somehow I remembered something wrongly...

File: components/script/dom/bindings/str.rs (hunk 23 above)
Comment (sendilkumarn) [inline components/script/dom/bindings/str.rs]: I have done the change, slightly modified. Now it works fine. 👍

File: components/script/dom/bindings/str.rs [hunk 27]
```diff
@@ -134,30 +134,36 @@ pub fn extended_filtering(tag: &str, range: &str) -> bool {
...
-                    Some(tag_subtag) => {
-                        // step 3c
-                        if range_subtag.eq_ignore_ascii_case(tag_subtag) {
-                            break;
+            match tag_iter.next() {
+                Some(tag_subtag) => {
+                    // step 3c
+                    if range_subtag.eq_ignore_ascii_case(tag_subtag) {
+                        current_tag_subtag = tag_iter.next();
+                        break;
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: Shouldn't this be `continue`?

File: components/script/dom/bindings/str.rs [hunk 28]
```diff
@@ -134,30 +134,36 @@ pub fn extended_filtering(tag: &str, range: &str) -> bool {
...
+                        // step 3d
+                        if tag_subtag.len() == 1 {
+                            return false;
                         } else {
-                            // step 3d
-                            if tag_subtag.len() == 1 {
+                            // else step 3e - continue with loop
+                            #[allow(unused_assignments)]
+                            current_tag_subtag = tag_iter.next();
+                            if current_tag_subtag == None {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: `if current_tag_subtag.is_none()`

File: components/script/dom/bindings/str.rs [hunk 29]
```diff
@@ -134,30 +134,36 @@ pub fn extended_filtering(tag: &str, range: &str) -> bool {
...
+                        break;
+                    } else {
+                        // step 3d
+                        if tag_subtag.len() == 1 {
+                            return false;
                         } else {
-                            // step 3d
-                            if tag_subtag.len() == 1 {
+                            // else step 3e - continue with loop
+                            #[allow(unused_assignments)]
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: I'm quite surprised that you need this annotation here.

File: components/script/dom/bindings/str.rs [hunk 30]
```diff
@@ -134,30 +134,36 @@ pub fn extended_filtering(tag: &str, range: &str) -> bool {
...
             if range_subtag.eq_ignore_ascii_case("*") {
                 continue;
             }
-            loop {
-                match tag_iter.next() {
-                    Some(tag_subtag) => {
-                        // step 3c
-                        if range_subtag.eq_ignore_ascii_case(tag_subtag) {
-                            break;
+            match tag_iter.next() {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: I thought we agreed that this shoud match against `current_tag_subtag.clone()` instead?

**Review by KiChjang — CHANGES_REQUESTED** [review]

**Comment by KiChjang** [comment]
@bors-servo r+

Thanks!

**Review by KiChjang — APPROVED** [review]

**Comment by sendilkumarn** [comment]
Thanks @KiChjang

**Comment by KiChjang** [comment]
```
  ▶ Unexpected subtest result in /dom/nodes/Element-matches.html:
  └ PASS [expected FAIL] In-document Element.matches: :lang pseudo-class selector, matching specified language with partial value (with no refNodes): #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/Element-matches.html:
  └ PASS [expected FAIL] Detached Element.matches: :lang pseudo-class selector, matching specified language with partial value (with no refNodes): #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/Element-matches.html:
  └ PASS [expected FAIL] Fragment Element.matches: :lang pseudo-class selector, matching specified language with partial value (with no refNodes): #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/Element-matches.html:
  └ PASS [expected FAIL] In-document Element.matches: :lang pseudo-class selector, matching specified language with partial value (1) (with no refNodes): #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All-xht.xht:
  └ PASS [expected FAIL] Document.querySelectorAll: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All-xht.xht:
  └ PASS [expected FAIL] Document.querySelector: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All-xht.xht:
  └ PASS [expected FAIL] Detached Element.querySelectorAll: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All-xht.xht:
  └ PASS [expected FAIL] Detached Element.querySelector: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All-xht.xht:
  └ PASS [expected FAIL] Fragment.querySelectorAll: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All-xht.xht:
  └ PASS [expected FAIL] Fragment.querySelector: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All-xht.xht:
  └ PASS [expected FAIL] In-document Element.querySelectorAll: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All-xht.xht:
  └ PASS [expected FAIL] In-document Element.querySelector: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All.html:
  └ PASS [expected FAIL] Document.querySelectorAll: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All.html:
  └ PASS [expected FAIL] Document.querySelector: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All.html:
  └ PASS [expected FAIL] Detached Element.querySelectorAll: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All.html:
  └ PASS [expected FAIL] Detached Element.querySelector: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All.html:
  └ PASS [expected FAIL] Fragment.querySelectorAll: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All.html:
  └ PASS [expected FAIL] Fragment.querySelector: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All.html:
  └ PASS [expected FAIL] In-document Element.querySelectorAll: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)

  ▶ Unexpected subtest result in /dom/nodes/ParentNode-querySelector-All.html:
  └ PASS [expected FAIL] In-document Element.querySelector: :lang pseudo-class selector, matching specified language with partial value: #pseudo-lang-div3:lang(en)
```
Great work! Now you'll need to update the test expectations, and after that you can r=me.
@bors-servo delegate+

**Comment by KiChjang** [comment]
Actually, let's also wait for http://build.servo.org/builders/linux-rel-css/builds/2695, since that is the buildbot running the CSS tests as well. I'm sure there are more test expectations to update... I'll keep you posted on when it's done.

**Comment by KiChjang** [comment]
Okay, false alarm, the CSS tests are all green. Just go ahead and update the test expectations for WPT and you can r=me.

**Comment by sendilkumarn** [comment]
r=me

**Comment by KiChjang** [comment]
Sorry, could you please squash your commits before we merge this?
